ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/api/v1/event/EventOps.scala

```diff
@@ -3,90 +3,89 @@
 import com.digitalasset.ledger.api.v1.transaction.TreeEvent.Kind.{
   Created => TreeCreated,
   Exercised => TreeExercised
 }
-import scalaz.Tag
+import com.digitalasset.ledger.api.v1.value.Identifier
 
 object EventOps {
 
   implicit class EventOps(val event: Event) extends AnyVal {
 
-    def eventId: EventId = event.event.eventId
+    def eventId: String = event.event.eventId
 
-    def witnesses: Seq[String] = event.event.witnesses
+    def witnessParties: Seq[String] = event.event.witnessParties
+    def witnessParties(set: Seq[String]): Event =
+      event.update(_.event.modify(_.witnessParties(set)))
+    def witnessParties(f: Seq[String] => Seq[String]): Event =
+      event.update(_.event.modify(_.witnessParties(f)))
 
-    def contractId: ContractId = event.event.contractId
+    def contractId: String = event.event.contractId
 
-    def templateId: String = event.event.templateId
-
-    def withWitnesses(witnesses: Seq[String]): Event = Event(event.event.withWitnesses(witnesses))
+    def templateId: Identifier = event.event.templateId
 
   }
 
   implicit class EventEventOps(val event: Event.Event) extends AnyVal {
 
-    def eventId: EventId = event match {
-      case Archived(value) => EventId(Ref.LedgerString.assertFromString(value.eventId))
-      case Created(value) => EventId(Ref.LedgerString.assertFromString(value.eventId))
+    def eventId: String = event match {
+      case Archived(value) => value.eventId
+      case Created(value) => value.eventId
       case Empty => throw new IllegalArgumentException("Cannot extract Event ID from Empty event.")
     }
 
-    def witnesses: Seq[String] = event match {
-      case c: Created => c.value.witnessParties
-      case a: Archived => a.value.witnessParties
+    def witnessParties: Seq[String] = event match {
+      case Archived(value) => value.witnessParties
+      case Created(value) => value.witnessParties
       case Empty => Seq.empty
     }
 
-    def templateId: String = event match {
-      case c: Created => c.templateId
-      case a: Archived => a.templateId
-      case Empty =>
-        throw new IllegalArgumentException("Cannot extract Template ID from Empty event.")
+    def witnessParties(set: Seq[String]): Event.Event = event match {
+      case a @ Archived(value) => a.copy(value.update(_.witnessParties := set))
+      case c @ Created(value) => c.copy(value.update(_.witnessParties := set))
+      case Empty => Empty
     }
 
-    def contractId: ContractId = {
-      val rawId = event match {
-        case Archived(value) => value.contractId
-        case Created(value) => value.contractId
-        case Empty =>
-          throw new IllegalArgumentException("Cannot extract contractId from Empty event.")
-      }
-      ContractId(Ref.LedgerString.assertFromString(rawId))
+    def witnessParties(f: Seq[String] => Seq[String]): Event.Event = event match {
+      case a @ Archived(value) => a.copy(value.update(_.witnessParties.modify(f)))
```

Comment:
Consider using
```scala
Archived(value.copy(witnessParties = f(value.witnessParties)))
```

Reply:
Addressed in https://github.com/digital-asset/daml/pull/4763/commits/084f83f6a7e9794c09cd508e4c4c0898039e0bea